File: design/swift/HousePour/HPToast.swift
// House Pour kit — HPToast (COMPONENTS.md).

import SwiftUI

public enum HPToastTone: Equatable { case neutral, good, bad }

public struct HPToastMessage: Equatable, Identifiable {
    public let id: UUID
    public let text: String
    public let tone: HPToastTone
    public init(_ text: String, tone: HPToastTone = .neutral) { id = UUID(); self.text = text; self.tone = tone }
    public static let autoDismiss: Double = 2.8
}

/// Inverted ink pill, fixed bottom centre, 26pt up. Fades; never slides.
public struct HPToast: View {
    let message: HPToastMessage
    public init(_ message: HPToastMessage) { self.message = message }
    private var lineColor: Color {
        switch message.tone {
        case .neutral: return HPTokens.Colors.toastLine
        case .good: return HPTokens.Colors.good.opacity(HPAlpha.toastTone)
        case .bad: return HPTokens.Colors.bad.opacity(HPAlpha.toastTone)
        }
    }
    public var body: some View {
        let shape = Capsule(style: .continuous)
        Text(message.text)
            .hpStyle(HPType.toast, color: HPTokens.Colors.toastText)
            .multilineTextAlignment(.center)
            .padding(.vertical, HPTokens.Space.buttonY)
            .padding(.horizontal, HPTokens.Space.topbarX)
            .background(shape.fill(HPTokens.Colors.toastBg))
            .hpBorder(shape, color: lineColor)
            .hpShadow(HPTokens.Shadow.toast, shape: shape, fill: HPTokens.Colors.toastBg)
            .accessibilityAddTraits(.isStaticText)
    }
    /// Distance from the bottom edge (the upstream `bottom: 26px`).
    public static let bottomOffset: CGFloat = HPTokens.Space.rowPad + HPTokens.Space.inputBottom
}

/// Hosts the toast over content, fading it in and out.
public struct HPToastHost: ViewModifier {
    @Binding var message: HPToastMessage?
    public func body(content: Content) -> some View {
        content.overlay(alignment: .bottom) {
            if let message {
                HPToast(message)
                    .padding(.bottom, HPToast.bottomOffset)
                    .padding(.horizontal, HPTokens.Space.columnSide)
                    .transition(.opacity)
                    .id(message.id)
            }
        }
        .animation(HPMotion.toast, value: message)
    }
}

public extension View {
    func hpToastHost(_ message: Binding<HPToastMessage?>) -> some View { modifier(HPToastHost(message: message)) }
}

File: design/kotlin/housepour/HPToast.kt
package ca.lucianlabs.housepour

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.asPaddingValues
import androidx.compose.foundation.layout.navigationBars
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay

enum class HPToastTone { NEUTRAL, GOOD, BAD }

data class HPToastMessage(val text: String, val tone: HPToastTone = HPToastTone.NEUTRAL, val id: Long = System.nanoTime())

@Stable
class HPToastState {
    var current: HPToastMessage? by mutableStateOf(null)
        private set

    fun show(text: String, tone: HPToastTone = HPToastTone.NEUTRAL) {
        current = HPToastMessage(text, tone)
    }

    fun dismiss(id: Long) {
        if (current?.id == id) current = null
    }
}

@Composable
fun rememberHPToastState(): HPToastState = remember { HPToastState() }

const val HP_TOAST_AUTO_DISMISS_MS = 2800L

/** The one dark surface. Fixed bottom centre 26pt up; fades `Motion.toast`; never slides; auto-dismisses after 2.8 s. */
@Composable
fun BoxScope.HPToastHost(state: HPToastState) {
    val message = state.current
    LaunchedEffect(message?.id) {
        val id = message?.id ?: return@LaunchedEffect
        delay(HP_TOAST_AUTO_DISMISS_MS)
        state.dismiss(id)
    }
    var shown by remember { mutableStateOf<HPToastMessage?>(null) }
    if (message != null) shown = message
    val nav = WindowInsets.navigationBars.asPaddingValues().calculateBottomPadding()
    AnimatedVisibility(
        visible = message != null,
        modifier = Modifier.align(Alignment.BottomCenter).padding(bottom = 26.dp + nav).padding(horizontal = HPTokens.Space.columnSide),
        enter = fadeIn(tween(HPTokens.Motion.TOAST_MS)),
        exit = fadeOut(tween(HPTokens.Motion.TOAST_MS)),
    ) {
        val m = shown ?: return@AnimatedVisibility
        HPToast(m)
    }
}

@Composable
fun HPToast(message: HPToastMessage, modifier: Modifier = Modifier) {
    val shape = RoundedCornerShape(HPTokens.Radius.pill)
    val line = when (message.tone) {
        HPToastTone.NEUTRAL -> HPTokens.Colors.toastLine
        HPToastTone.GOOD -> HPTokens.Colors.good
        HPToastTone.BAD -> HPTokens.Colors.bad
    }
    Box(
        modifier = modifier
            .widthIn(max = HPTokens.Space.columnMax)
            .hpShadow(HPTokens.Radius.pill, HPTokens.Shadow.toast)
            .background(HPTokens.Colors.toastBg, shape)
            .border(HPTokens.BORDER_WIDTH.dp, line, shape)
            .padding(horizontal = HPTokens.Space.buttonX, vertical = HPTokens.Space.inputY)
            .semantics { liveRegion = LiveRegionMode.Polite },
    ) {
        HPText(message.text, HPTokens.Type.toast, HPTokens.Colors.toastText, textAlign = TextAlign.Center)
    }
}
